Sources/Filters/General/ContourTriangulator/index.js

```diff
@@ -0,0 +1,269 @@
+  vtkCCSSplitAtPinchPoints(
+    newPolys,
+    points,
+    polyGroups,
+    polyEdges,
+    normal,
+    oriented
+  );
+
+  // ------ Triangulation code ------
+
+  // Go through all polys and triangulate them
+  for (let polyId = 0; polyId < polyGroups.length; polyId++) {
+    // If group is empty, then poly was a hole without a containing poly
+    if (polyGroups[polyId].length === 0) {
+      // eslint-disable-next-line no-continue
+      continue;
+    }
+
+    if (
+      !vtkCCSTriangulate(
+        newPolys[polyId],
+        points,
+        polyEdges[polyId],
+        originalEdges,
+        polys,
+        normal
+      )
+    ) {
+      triangulationFailure = false;
+      // Diagnostic code: show the polys as outlines
+      // const lines = polyData.getLines();
+      // const poly = newPolys[polyId];
+      // lines.insertNextCell(poly.length + 1);
+      // for (let jjj = 0; jjj < poly.length; jjj++) {
+      //   lines.insertCellPoint(poly[jjj]);
+      // }
+      // lines.insertCellPoint(poly[0]);
+    }
+  }
+
+  return !triangulationFailure;
+}
+
+// ---------------------------------------------------
+function triangulatePolygon(polygon, points, triangles) {
+  const poly = [...polygon];
+  const polys = [poly];
+
+  const originalEdges = [];
+  const polyEdges = [];
+  vtkCCSFindTrueEdges(polys, points, polyEdges, originalEdges);
+  const edges = polyEdges[0];
+
+  let success = true;
+  const normal = [];
+  const norm = vtkPolygon.getNormal(poly, points, normal);
+  if (norm !== 0) {
+    success = vtkCCSTriangulate(
+      poly,
+      points,
+      edges,
+      originalEdges,
+      triangles,
+      normal
+    );
+  }
+  return success;
+}
+
+export const STATIC = { triangulateContours, triangulatePolygon };
+
+function vtkContourTriangulator(publicAPI, model) {
+  // Set our className
+  model.classHierarchy.push('vtkContourTriangulator');
+
+  publicAPI.requestData = (inData, outData) => {
+    // implement requestData
+    const input = inData[0];
+    const output = vtkPolyData.newInstance();
+    outData[0] = output;
+
+    if (!input) {
+      vtkErrorMacro('Invalid or missing input');
+      return false;
+    }
+
+    let triangulationError = false;
+
+    const lines = input.getLines();
+    if (lines == null || lines.getNumberOfCells() === 0) {
+      return true;
+    }
+
+    input.buildCells();
+
+    const polysArray = vtkCellArray.newInstance({
+      dataType: VtkDataTypes.DOUBLE,
+      empty: true,
+    });
+    output.setPolys(polysArray);
+    output.setPoints(input.getPoints());
+    output.getPointData().passData(input.getPointData());
+
+    triangulationError = !triangulateContours(
+      input,
+      input.getNumberOfVerts(),
+      lines.getNumberOfCells(),
+      polysArray,
+      []
+    );
+
+    if (triangulationError && model.triangulationErrorDisplay) {
+      vtkErrorMacro('Triangulation failed, output might have holes.');
```

Comment:
if triangulation fails, then a safer approach is to NOT triangularize, right ?

Then how about a `triangulate` option instead that calls `triangulateContours` if ON.

Reply:
I am not sure I am following. The purpose of `vtkContourTriangulator` is to `triangulate`. So why have the option to set that to OFF? What would we expect the algorithm to do in that case?

Reply:
Ah I see what you mean

Reply:
We would need to add an else in that case that adds the polygons to the output, instead of triangles. Should I proceed with that?